clingo: use python from spec

severinstrobl opened the pull request:
Without explicitly providing the path to the Python executable, clingo is prone to pick up the wrong Python package. As a safeguard, the required version is set explicitly based on the spec.

severinstrobl committed:
clingo: use python from spec

Without explicitly providing the path to the Python executable, clingo
is prone to pick up the wrong Python package. As a safeguard, the
required version is set explicitly based on the spec.

cosmicexplorer commented:
This should probably go in before #20159, as it is more correct than the corresponding change there:
https://github.com/spack/spack/blob/c49a3047912bdfa8465c6c4cbbec2b66482a1615/var/spack/repos/builtin/packages/clingo/package.py#L77-L93

adamjstewart reviewed (changes requested):
Python is not in the spec unless `+python`, you need an if-statement to check this before adding the flags.

severinstrobl commented:
Good point about the `+python`, somehow I overlooked that Python is (somewhat?) optional. Unless I'm mistaken, also the original version does not handle this correctly, as all the other CMake arguments regarding the Python bindings do not consider the variant at all. Do you want me to fix this in this merge request as well or should this be addressed separately?

adamjstewart commented:
Please fix it in this PR.

severinstrobl committed:
clingo: handle python variant in cmake phase

Clingo can be configured without Python bindings, yet the CMake
configure step has to explicitly handle this. Additionally, for CMake
versions <= 3.15.0 the path to the Python binary is set explicitly based
on the spec to ensure CMake picks up the correct Python installation.

severinstrobl commented:
The support for the option `CLINGO_PYTHON_VERSION` is only available in the master branch, not any of the releases. Is it acceptable for a package to specify CMake options based on the check `self.version == Version('master')` or is this considered bad practice?

adamjstewart commented:
That's fine, as long as it's reasonable to expect that it will be present in future versions. `master` is a moving target, so we don't want to get too specific about what flags it can and can't accept. If you know what the next feature release number is going to be, you can also say `@5.5:` or something like that.

adamjstewart commented on var/spack/repos/builtin/packages/clingo/package.py:
```diff
@@ -56,11 +56,35 @@ def cmake_args(self):
+            args += [
+                '-DCLINGO_BUILD_WITH_PYTHON=ON',
+                '-DCLINGO_REQUIRE_PYTHON=ON',
+                '-DCLINGO_BUILD_PY_SHARED=ON',
```
What does this flag control? Does it depend on whether `'+shared' in spec['python']`?

severinstrobl replied:
Good question, unfortunately I'm not that familiar with the internals of the library, maybe the original author of this package can shed some light. There seems to be an additional library `libpyclingo` that uses this flag. As far as I can tell, this library is not used by the Python module, which links against `libclingo.so` directly.

adamjstewart replied:
Alright, maybe @tgamblin knows

severinstrobl committed:
clingo: enforce python path for all cmake versions

Explicitly specifying the path to the Python executable to use is
necessary independent of the CMake version to avoid issues with multiple
Python versions.

severinstrobl commented:
With the recent changes to the package when using CMake >= 3.16.0 this is not an issue any more, so I'll close this one.
